### PR TITLE
Or react redux

### PR DESCRIPTION
**Pull request by oarozu**
Finger Exercise 2

### CODE REVIEW

File: react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js [hunk 1]
```diff
@@ -8,19 +9,47 @@ const initialState = {
...
-    case actions.ADD_ITEM: // TODO to implement the logic
-      return { ...state };
-    case actions.REMOVE_ITEM: // TODO to implement the logic
-      return { ...state };
-    case actions.SEARCH_ITEM: // TODO to implement the logic
-      return { ...state };
+    case actions.GET_BOOKS:
+      return {
+        ...state,
+        books: state.books.concat(DATA),
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js]: getBooks action already have DATA as a paylod, so you can only write books: action.payload and so on

File: react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js [hunk 2]
```diff
@@ -8,19 +9,47 @@ const initialState = {
...
+        originalData: state.originalData.concat(DATA)
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js]: same here

File: react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js [hunk 3]
```diff
@@ -8,19 +9,47 @@ const initialState = {
...
+      return {
+        ...state,
+        books: filterByName(state.originalData, action.value)
+      };
     default:
       return state;
   }
 }
 
+function filterByName(arr, value) {
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js]: export in utils file

File: react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js [hunk 4]
```diff
@@ -8,19 +9,47 @@ const initialState = {
...
+  return arr.filter(function (a) {
+    return a.name.toLowerCase().indexOf(value.toLowerCase()) !== -1;
+  })
+}
+
+function addItem(selected, id) {
+  var index = selected.findIndex(i => i.id === id);
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/redux/book/reducer.js]: same

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 5]
```diff
@@ -2,20 +2,16 @@ import React, { PureComponent, Fragment } from 'react';
...
-  };
 
-  toggleContent = () => {
-    this.setState(prevState => ({
-      open: !prevState.open
-    }));
-  };
+  // state = {
+  //   open: false
+  // }
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: If you arent going to use these, please delete it

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 6]
```diff
@@ -28,10 +24,10 @@ class ShoppingCart extends PureComponent {
     const { data } = this.props;
     return (
       <Fragment>
-        <Button className={styles.buttonCart} onClick={this.toggleContent}>
+        <Button className={styles.buttonCart} onClick={this.props.toggleContent}>
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: destruct props please!

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 7]
```diff
@@ -47,4 +43,25 @@ ShoppingCart.propTypes = {
   removeItem: func.isRequired
 };
 
-export default ShoppingCart;
+const mapStateToProps = state => {
+  return {
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: You dont have to write return in arrow functions

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 8]
```diff
@@ -47,4 +43,25 @@ ShoppingCart.propTypes = {
...
+    open: state.plays.open,
+    bookSelected: state.books.bookSelected
+  };
+};
+
+const mapDispatchToProps = dispatch => {
+  return {
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: Same

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 9]
```diff
@@ -47,4 +43,25 @@ ShoppingCart.propTypes = {
...
+    removeItem: (itemId) => {
+      dispatch({type: '@@BOOK/REMOVE_ITEM', itemId});
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: You should dispatch the action with actionCreators.actionName and send itemId

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 10]
```diff
@@ -47,4 +43,25 @@ ShoppingCart.propTypes = {
...
+    },
+    addItem: (itemId) =>  {
+      dispatch({type: '@@BOOK/ADD_ITEM', itemId});
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: same here

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 11]
```diff
@@ -47,4 +43,25 @@ ShoppingCart.propTypes = {
...
+    },
+    toggleContent: () => {
+      dispatch({type: '@@CART/TOGGLE_CONTENT'});
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: same here

File: react/finger-exercises/redux/02 - react-redux/src/index.js [hunk 12]
```diff
@@ -1,9 +1,16 @@
...
 import './scss/index.scss';
 import registerServiceWorker from './registerServiceWorker';
 
-ReactDOM.render(<App />, document.getElementById('root'));
+ReactDOM.render(
+  <Provider store={store}>
+    <App />
+  </Provider>,
+  document.getElementById('root'));
+  console.log(store.getState());
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/index.js]: delete console.log please!

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js (hunk 11 above)
Comment (mariadelacamara) [reply react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: You can maintain state open and toggleContent as a local state

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js [hunk 13]
```diff
@@ -9,20 +9,22 @@ class Item extends PureComponent {
...
 
   removeItem = () => {
     const { item, removeItem } = this.props;
     removeItem(item.id);
+    console.log(item.id);
   };
 
   render() {
-    const { item } = this.props;
+    console.log(this.props);
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js]: please delete console.log :)

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js [hunk 14]
```diff
@@ -9,20 +9,22 @@ class Item extends PureComponent {
   addItem = () => {
     const { item, addItem } = this.props;
     addItem(item.id);
-  };
+    };
 
   removeItem = () => {
     const { item, removeItem } = this.props;
     removeItem(item.id);
+    console.log(item.id);
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js]: please delete console.log

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 15]
```diff
@@ -20,21 +23,21 @@ class ShoppingCart extends PureComponent {
   total = (accumulator, currentValue) => accumulator + currentValue.quantity;
 
   renderItem = item => {
+    console.log(actionsCreators);
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: same!

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js [hunk 16]
```diff
@@ -20,21 +23,21 @@ class ShoppingCart extends PureComponent {
...
     return (
       <Fragment>
         <Button className={styles.buttonCart} onClick={this.toggleContent}>
           <i className="fa fa-shopping-cart" />
         </Button>
         <div className={`${styles.container} ${this.state.open ? styles.open : ''}`}>
           <h1 className={styles.title}>Cart</h1>
-          <ul className={styles.content}>{data.map(this.renderItem)}</ul>
-          <h2 className={`${styles.title} ${styles.total}`}>Total: {data.reduce(this.total, 0)}</h2>
+          <ul className={styles.content}>{this.props.bookSelected.map(this.renderItem)}</ul>
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/index.js]: destruct props

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js [hunk 17]
```diff
@@ -1,53 +1,35 @@
...
 import ShoppingCart from './components/ShoppingCart';
 import styles from './styles.scss';
+import { connect } from 'react-redux';
 
 class App extends Component {
-  state = {
-    books: [],
-    bookSelected: []
-  };
+  props: any;
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js]: this isn't necesary

File: react/finger-exercises/redux/02 - react-redux/src/utils/utils.js [hunk 18]
```diff
@@ -0,0 +1,5 @@
+export default function filterByName(arr, value) {
+  return arr.filter(function (a) {
+    return a.name.toLowerCase().indexOf(value.toLowerCase()) !== -1;
+  })
+}
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/utils/utils.js]: add empty line at the end of file

File: react/finger-exercises/redux/02 - react-redux/src/utils/utils.js [hunk 19]
```diff
@@ -0,0 +1,5 @@
+export default function filterByName(arr, value) {
+  return arr.filter(function (a) {
+    return a.name.toLowerCase().indexOf(value.toLowerCase()) !== -1;
```
Comment (mariadelacamara) [inline react/finger-exercises/redux/02 - react-redux/src/utils/utils.js]: you can write this as an arrow function
arr.filter(a => a.name ...)

Besides, I would write a more descriptive name of each element

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js [hunk 20]
```diff
@@ -34,11 +32,9 @@ class Item extends PureComponent {
     );
   }
 }
-
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js]: Don't remove this empty line

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js [hunk 21]
```diff
@@ -34,11 +32,9 @@ class Item extends PureComponent {
...
 Item.propTypes = {
   item: bookSelectedPropType,
   addItem: func.isRequired,
   removeItem: func.isRequired
 };
-
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/components/ShoppingCart/components/Item/index.js]: Same here

File: react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js [hunk 22]
```diff
@@ -57,22 +38,34 @@ class App extends Component {
...
-          )}
+              <div className={styles.noData}>
+                <h2 className={styles.title}>No Data</h2>
+              </div>
+            )}
         </div>
-        {this.state.bookSelected.length ? (
-          <ShoppingCart data={this.state.bookSelected} addItem={this.addItem} removeItem={this.removeItem} />
+        {this.props.bookSelected.length ? (
+          <ShoppingCart data={this.props.bookSelected} addItem={this.props.addItem} removeItem={this.props.removeItem} />
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/app/screens/App/index.js]: Destructuring the props in this method

File: react/finger-exercises/redux/02 - react-redux/src/redux/plays/actions.js [hunk 23]
```diff
@@ -0,0 +1,12 @@
+export const actions = {
+  TOGGLE_CONTENT: '@@CART/TOGGLE_CONTENT'
+};
+
+const actionsCreators = {
+  toggleContent: () => ({
+    type: actions.TOGGLE_CONTENT,
```
Comment (erikankodelia) [inline react/finger-exercises/redux/02 - react-redux/src/redux/plays/actions.js]: Are you using this action? If not, remove it.

**Pull request merged**